File: gitdot-auth/src/handler/device/poll_token.rs
use axum::{Json, extract::State, http::StatusCode};

use gitdot_api::endpoint::auth::device::poll_token as api;
use gitdot_core::dto::PollTokenRequest;

use crate::{
    app::{AppError, AppResponse, AppState},
    dto::IntoApi,
};

pub async fn poll_token(
    State(state): State<AppState>,
    Json(body): Json<api::PollTokenRequest>,
) -> Result<AppResponse<api::PollTokenResponse>, AppError> {
    let request = PollTokenRequest {
        device_code: body.device_code,
        client_id: body.client_id,
    };
    state
        .device_service
        .poll_token(request)
        .await
        .map_err(AppError::from)
        .map(|token| AppResponse::new(StatusCode::CREATED, token.into_api()))
}

File: gitdot-web/packages/gitdot-api/src/endpoint/auth/device/poll-token.ts
import { z } from "zod";
import { TokenResource } from "../../../resource";
import type { Endpoint } from "../../endpoint";

export const PollTokenRequest = z.object({
  device_code: z.string(),
  client_id: z.string(),
});
export type PollTokenRequest = z.infer<typeof PollTokenRequest>;

export const PollTokenResponse = TokenResource;
export type PollTokenResponse = z.infer<typeof PollTokenResponse>;

export const PollToken = {
  path: "/auth/device/token",
  method: "POST",
  request: PollTokenRequest,
  response: PollTokenResponse,
} as const satisfies Endpoint;
export type PollToken = typeof PollToken;
